beacon_node/beacon_chain/src/validator_custody.rs

```diff
@@ -0,0 +1,433 @@
+    /// that is attached to this node.
+    ///
+    /// This is the number that we use to compute the cgc value that
+    /// we advertise to our peers in the metadata and enr values.
+    advertised_validator_custody_count: AtomicU64,
+    /// This is the validator custody count that we use to compute the number of columns we need to
+    /// custody at head (while syncing or when receiving gossip).
+    validator_custody_count_at_head: AtomicU64,
+    /// Is the node run as a supernode based on current cli parameters.
+    pub current_is_supernode: bool,
+    /// The persisted value for `is_supernode` based on the previous run of this node.
+    ///
+    /// Note: We require this value because if a user restarts the node with a higher cli custody
+    /// count value than in the previous run, then we should continue advertising the custody
+    /// count based on the old value than the new one since we haven't backfilled the required
+    /// columns.
+    persisted_is_supernode: bool,
+    /// Maintains all the validators that this node is connected to currently
+    validator_registrations: RwLock<ValidatorRegistrations>,
+    sender: Sender<CustodyContextMessage>,
+}
+
+impl CustodyContext {
+    /// Create a new custody default custody context object when no persisted object
+    /// exists.
+    ///
+    /// The `is_supernode` value is based on current cli parameters.
+    pub fn new(is_supernode: bool) -> Self {
+        let (sender, _) = channel(CHANNEL_CAPACITY);
+
+        Self {
+            advertised_validator_custody_count: AtomicU64::new(0),
+            validator_custody_count_at_head: AtomicU64::new(0),
+            current_is_supernode: is_supernode,
+            persisted_is_supernode: is_supernode,
+            validator_registrations: Default::default(),
+            sender,
+        }
+    }
+
+    /// Deserialize a `CustodyContext` from SSZ bytes.
+    pub fn new_from_persisted_custody_context(
+        ssz_context: CustodyContextSsz,
+        is_supernode: bool,
+    ) -> Self {
+        let (sender, _) = channel(CHANNEL_CAPACITY);
+        CustodyContext {
+            advertised_validator_custody_count: AtomicU64::new(
+                ssz_context.advertised_validator_custody,
+            ),
+            validator_custody_count_at_head: AtomicU64::new(ssz_context.validator_custody_at_head),
+            current_is_supernode: is_supernode,
+            persisted_is_supernode: ssz_context.persisted_is_supernode,
+            validator_registrations: Default::default(),
+            sender,
+        }
+    }
+
+    pub fn subscribe(&self) -> Receiver<CustodyContextMessage> {
+        self.sender.subscribe()
+    }
+
+    /// Register a new validator index and updates the list of validators if required.
+    ///
+    /// Also modifies the internal structures if the validator custody has changed to
+    /// update the `custody_column_count`.
+    pub fn register_validator<E: EthSpec>(
```

Comment:
perhaps rename this to `register_validators` given we're accepting a `Vec`

Reply:
Addressed in https://github.com/sigp/lighthouse/pull/7578/commits/1adcc4a5406911fa8677a51507a787a9c1565e2c